### PR TITLE
`Programming exercises`: Change the solution entry and testwise coverage editors to Monaco

### PR DESCRIPTION
**Pull request by pzdr7**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

We want to replace Ace in Artemis with Monaco. The (presumably unused?) Code Hints feature relies on Ace to display & edit code snippets.

### Description

- Introduced Monaco into the solution entry modals and editor
- Minor layout changes (moved the save/submit button)
- The syntax highlighting of the code now depends on the file extension (instead of always being java)

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming exercise

1. Log in to Artemis
2. Navigate to course management > your course > exercises > your programming exercise
3. Click "Hints". It should show you "No Exercise Hint found."
4. Click "Generate code hints" (**Note**: You have to wait for the solution repository build to finish before doing this.)
5. Click "Generate structural snippets"
6. For one of the generated entries, click "View" and verify that the modal looks good.
7. For one of the generated entries, click "Edit" and verify that the modal looks good. Make a change and save it, then verify that the change was actually saved.
8. Click "Create manual fragment". Select a file and a test case, write the fragment and submit it. Verify in the list that your changes were saved correctly.
9. Click "Next"
10. Click "Generate Code Hints"
11. Expand one of the code hints and check that it looks good.
12. Edit one of the code hints and check that it looks good (**Note**: To save the code hint, you may have to enter a valid description at the top.). Verify that your changes are saved

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| manual-solution-entry-creation-modal.component.ts | 90.9% | ✅ |
| solution-entry.component.ts | 100% | ✅ |
| monaco-editor.component.ts | 92% | ✅ |

### Screenshots

<details>
  <summary>Modals</summary>

image

image

image

</details>

<details>
  <summary>Code hints + editing</summary>

image

image

</details>

## Summary by CodeRabbit

- **New Features**
	- Enhanced the manual solution entry modal with dynamic file selection updates and improved layout.
	- Integrated Monaco Editor for a more advanced code editing experience.
	- Added methods to manage content height and specify starting line numbers in the Monaco Editor.

- **Bug Fixes**
	- Improved test coverage for the Monaco Editor's start line number functionality.

- **Documentation**
	- Updated test cases to reflect the new editor technology and better verify component behaviors.

### CODE REVIEW

**Review by edkaya — APPROVED** [review]
Tested on TS2, works as described 👍

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by BBesrour — DISMISSED** [review]
code

**Review by Jan-Thurner — APPROVED** [review]
Tested on TS1. Works as described.

File: src/main/webapp/app/shared/monaco-editor/monaco-editor.component.ts [hunk 1]
```diff
@@ -109,7 +109,7 @@ export class MonacoEditorComponent implements OnInit, OnDestroy {
         });
         resizeObserver.observe(this.monacoEditorContainerElement);
 
-        this.textChangedListener = this._editor.onDidChangeModelContent(() => {
+        this._editor.onDidChangeModelContent(() => {
```
Comment (pzdr7) [inline src/main/webapp/app/shared/monaco-editor/monaco-editor.component.ts:112]:
```suggestion
        this.textChangedListener = this._editor.onDidChangeModelContent(() => {
```

**Review by edkaya — APPROVED** [review]
Reapprove after merge conflict

**Review by JohannesStoehr — APPROVED** [review]
Reapprove code after merge conflict

**Review by BBesrour — DISMISSED** [review]
Reapprove code after merge conflict

**Review by Jan-Thurner — DISMISSED** [review]
reprove after merge conflict

**Pull request merged**